src/components/KTreeList/KTreeList.vue

```diff
@@ -0,0 +1,419 @@
+      v-bind="draggableAttrs"
+      :disabled="disableDrag"
+      :list="internalList"
+      :group="{ name: 'k-tree-list', put: !maxLevelReached }"
+      :move="checkMove"
+      :level="level"
+      @start="onStartDrag"
+      @end="onStopDrag"
+      @change="handleChangeEvent"
+    >
+      <template #item="{ element }">
+        <div
+          :style="indentStyle"
+          :class="{
+            'non-root': level !== 0
+          }"
+          class="k-tree-item-container"
+        >
+          <KTreeItem
+            :key="`tree-item-${key}`"
+            :item="element"
+            :disabled="disableDrag"
+            :class="{
+              'has-children': hasChildren(element)
+            }"
+            @selected="handleSelection"
+          >
+            <template #item-icon>
+              <slot
+                :item="element"
+                name="item-icon"
+              >
+                <KIcon
+                  v-if="element.icon !== 'none'"
+                  :icon="element.icon ? element.icon : 'treeDoc'"
+                  :secondary-color="iconSecondaryColor(element)"
+                  size="24"
+                />
+              </slot>
+            </template>
+            <template #item-label>
+              <slot
+                :item="element"
+                name="item-label"
+              >
+                {{ element.name }}
+              </slot>
+            </template>
+          </KTreeItem>
+          <KTreeList
+            :key="`tree-item-children-${key}`"
+            v-model="element.children"
+            :level="level + 1"
+            :max-level="maxLevel"
+            :disable-drag="disableDrag"
+            :parent-id="element.id"
+            @change="handleChangeEvent"
+            @selected="handleSelection"
+          >
+            <!-- if TS errors appear on these lines switch to `v-slot:` notation and Save to let linter clear -->
+            <template #[itemIcon]="slotProps">
+              <slot
+                v-bind="slotProps"
+                name="item-icon"
+              />
+            </template>
+            <template #[itemLabel]="slotProps">
+              <slot
+                v-bind="slotProps"
+                name="item-label"
+              />
+            </template>
+          </KTreeList>
+        </div>
+      </template>
+    </draggable>
+  </div>
+</template>
+
+<script lang="ts" setup>
+import { defineProps, defineEmits, computed, ref, watch, onMounted, PropType } from 'vue'
+import draggable from 'vuedraggable'
+import KTreeItem, { TreeListItem } from '@/components/KTreeList/KTreeItem.vue'
+
+export interface ChangeEvent {
+  items: TreeListItem[]
+}
+
+export interface ChildChangeEvent {
+  parent: string,
+  children: TreeListItem[]
+}
+
+const props = defineProps({
+  modelValue: {
+    type: Array as PropType<TreeListItem[]>,
+    default: () => [],
+  },
+  items: {
+    type: Array as PropType<TreeListItem[]>,
+    default: null,
```

Comment:
Since you're saying `item.id` is required (along with the other required params) you need to add a validator here to ensure the properties are set accordingly.

Reply:
Ok, I thought I could get it for free since it's required in the type interface 😞